rust-runtime/aws-smithy-checksums/src/http.rs

```diff
@@ -0,0 +1,226 @@
+        let hash = self.state.unwrap_or_default();
+        HeaderValue::from_str(&base64::encode(u32::to_be_bytes(hash)))
+            .expect("will always produce a valid header value from a CRC32C checksum")
+    }
+}
+
+impl HttpChecksum for Sha1 {
+    fn header_name(&self) -> HeaderName {
+        SHA_1_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        use sha1::Digest;
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.finalize();
+        HeaderValue::from_str(&base64::encode(&hash[..]))
+            .expect("will always produce a valid header value from a SHA-1 checksum")
+    }
+}
+
+impl HttpChecksum for Sha256 {
+    fn header_name(&self) -> HeaderName {
+        SHA_256_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        use sha2::Digest;
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.finalize();
+        HeaderValue::from_str(&base64::encode(&hash[..]))
+            .expect("will always produce a valid header value from a SHA-256 checksum")
+    }
+}
+
+impl HttpChecksum for Md5 {
+    fn header_name(&self) -> HeaderName {
+        MD5_HEADER_NAME.clone()
+    }
+
+    fn header_value(self: Box<Self>) -> HeaderValue {
+        use md5::Digest;
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.finalize();
+        HeaderValue::from_str(&base64::encode(&hash[..]))
+            .expect("will always produce a valid header value from an MD5 checksum")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        new_from_algorithm, HttpChecksum, CRC_32_C_NAME, CRC_32_NAME, SHA_1_NAME, SHA_256_NAME,
+    };
+
+    #[test]
+    fn test_trailer_length_of_crc32_checksum_body() {
+        let checksum = new_from_algorithm(CRC_32_NAME);
+        let expected_size = 29;
+        let actual_size = HttpChecksum::size(&*checksum);
+        assert_eq!(expected_size, actual_size)
+    }
+
+    #[test]
+    fn test_trailer_length_of_crc32c_checksum_body() {
+        let checksum = new_from_algorithm(CRC_32_C_NAME);
+        let expected_size = 30;
+        let actual_size = HttpChecksum::size(&*checksum);
+        assert_eq!(expected_size, actual_size)
+    }
+
+    #[test]
+    fn test_trailer_length_of_sha1_checksum_body() {
+        let checksum = new_from_algorithm(SHA_1_NAME);
+        let expected_size = 48;
+        let actual_size = HttpChecksum::size(&*checksum);
+        assert_eq!(expected_size, actual_size)
+    }
+
+    #[test]
+    fn test_trailer_length_of_sha256_checksum_body() {
+        let checksum = new_from_algorithm(SHA_256_NAME);
+        let expected_size = 66;
+        let actual_size = HttpChecksum::size(&*checksum);
+        assert_eq!(expected_size, actual_size)
```

Comment:
should we have some tests of actual checksums?

imagining bugs like typoing the checksum algorithm